server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java

```diff
@@ -868,4 +876,76 @@ public void onFailure(Exception e) {
             )
         );
     }
+
+    private final ConsistencyChecker consistencyChecker = new ConsistencyChecker();
+
+    /**
+     * After receiving an ack from the master confirming that it marked a shard snapshot as complete, checks the local cluster state to
+     * confirm that it was updated to reflect this and logs a message if not. Not 100% watertight because this node might be lagging behind
+     * and hasn't received the cluster state yet, or else it might have been removed from the cluster, but both of these things will also
+     * be visible in the logs.
+     */
+    // visible for testing
+    class ConsistencyChecker {
+
+        // dedicated logger so we can separate these logs from other SnapshotShardsService DEBUG logs
+        private static final Logger CONSISTENCY_CHECKER_LOGGER = LogManager.getLogger(ConsistencyChecker.class);
+
+        private record CheckTask(Snapshot snapshot, ShardId shardId) {}
+
+        private final AtomicInteger queuedTaskCount = new AtomicInteger(0); // atomic to ensure max one thread processing the queue
+        private final Queue<CheckTask> queue = new ConcurrentLinkedQueue<>();
+
+        void ensureShardComplete(Snapshot snapshot, ShardId shardId) {
+            if (CONSISTENCY_CHECKER_LOGGER.isDebugEnabled() == false) {
+                return;
+            }
+
+            if (queuedTaskCount.get() > 1000) {
+                // racy check, we only need an approximate limit
+                return;
+            }
+
+            queue.add(new CheckTask(snapshot, shardId));
+            if (queuedTaskCount.getAndIncrement() == 0) {
+                threadPool.generic().execute(this::runCheck);
+            } // else a runCheck is already running somewhere and will pick up the task we just added
+        }
+
+        private void runCheck() {
```

Comment:
```suggestion
        /**
         * Empties the {@link #queue} and verifies that the cluster state shows a successful shard snapshot for each {@link CheckTask}.
         * A debug level message is logged if the cluster state does not show a completed shard snapshot: this could be a bug, or a race
         * with a cluster state update not being applied yet.
         */
        private void runCheck() {
```